### PR TITLE
Add IRQ_OFFLOAD_NESTED, with a portable (!) nested interrupt test

### PR DESCRIPTION
**Pull request by andyross**
Nested interrupts on xtensa weren't being exercised due to bitrot (I wrote a test long ago that fell out of the suite, and the nested interrupt test was always arch-specific and never got a port).  This extends the irq_offload() API to be able to interrupt an ISR on supported architectures (just x86 and xtensa for now), and adds a trivially small case to tests/kernel/common to validate it.

### CODE REVIEW

**Comment by hongshui3000** [comment]
> Nested interrupts on xtensa weren't being exercised due to bitrot (I wrote a test long ago that fell out of the suite, and the nested interrupt test was always arch-specific and never got a port). This extends the irq_offset() API to be able to interrupt an ISR on supported architectures (just x86 and xtensa for now), and adds a trivially small case to tests/kernel/common to validate it.

This test seems to have an obvious place to pay attention to. He assumes that the sw interrupt has a higher priority than the time0 interrupt? When this condition is not met, I don't think the test will work as it should?
For example, my architecture should need to be adjusted, and these two interrupts cannot be used directly to test interrupt nesting
image

**Comment by andyross** [comment]
Yes, if you don't have a software interrupt at higher priority than your lowest timer this test won't work, and the board should deselect IRQ_OFFSET_NESTED.  I wish it were true you could rely on the highest software interrupt being higher than the lowest timer, but Cadence lets people configure cores in all sorts of ways.

**Review by dcpleung — APPROVED** [review]

**Comment by andyross** [comment]
Add some docs to explain the edge case called out above (being nested-capable isn't magic, it still requires that you be running at an interruptible level), clean up one stray newline-at-EOF.  Seems like CI had glitched on this one anyway and was showing a lot of download failures.

File: tests/kernel/common/src/irq_offload.c [hunk 1]
```diff
@@ -199,3 +199,46 @@ __no_optimization void test_nop(void)
...
+	if (!IS_ENABLED(CONFIG_IRQ_OFFLOAD_NESTED)) {
+		ztest_test_skip();
+	}
+
+	k_timer_init(&nestoff_timer, nestoff_timer_fn, NULL);
+
+	zassert_false(timer_executed, "timer ran too soon");
+	zassert_false(nested_executed, "nested irq_offload ran too soon");
+
+	k_timer_start(&nestoff_timer, K_MSEC(1000), K_FOREVER);
```
Comment (andyross) [inline tests/kernel/common/src/irq_offload.c:239]: Whoops, this was supposed to be K_TICKS(1), I bumped the time to be able to watch the logging and forgot to tune it back down before pushing.  Will fix.

**Review by sylvioalves — APPROVED** [review]
@andyross Looks good to me. I tested in ESP32 and ESP32S2, no issues.
In fact, this fixes and old issue regarding fast GPIO interrupt toggling when `CONFIG_ASSERT=y`. It would assert because of `arch_is_in_isr()` check. Now it does not get asserted anymore.

**Pull request merged**